pymc_marketing/mmm/models/components/base.py

```diff
@@ -0,0 +1,155 @@
+import warnings
+from inspect import signature
+from typing import Any
+
+from pytensor import tensor as pt
+
+from pymc_marketing.mmm.utils import _get_distribution
+
+
+class ParameterPriorException(Exception):
+    """Error when the functions and specified priors don't match up."""
+
+    def __init__(self, priors: set[str], parameters: set[str]) -> None:
+        self.priors = priors
+        self.parameters = parameters
+
+        msg = "The function parameters and priors don't line up."
+
+        if self.priors:
+            msg = f"{msg} Missing prior: {self.priors}."
+
+        if self.parameters:
+            msg = f"{msg} Missing parameter: {self.parameters}."
+
+        super().__init__(msg)
+
+
+class Transformation:
+    """Base class for adstock and saturation functions.
+
+    The subclasses will need to implement the following attributes:
+
+    - function: The function that will be applied to the data.
+    - prefix: The prefix for the variables that will be created.
+    - default_priors: The default priors for the parameters of the function.
+
+    In order to make a new saturation or adstock function, this subclass would
+    not be used. Instead, the subclasses would be used.
+
+    """
+
+    prefix: str
+    default_priors: dict[str, Any]
+    function: Any
+
+    def __init__(self, priors: dict | None = None) -> None:
+        self._checks()
+
+        priors = priors or {}
+        self.function_priors = {**self.default_priors, **priors}
```

Comment:
Thoughts on being able to change the prefix as well? 
```suggestion
    def __init__(self, priors: dict | None = None, prefix: str | None = None) -> None:
        self._checks()

        priors = priors or {}
        self.function_priors = {**self.default_priors, **priors}
        self.prefix = prefix or self.prefix
```
This would only affect the subclasses which would still have defaults. However, this would provide ability to use multiple when used in a custom MMM.

For instance, using menten and logistic in the same model
```python
menten = MichaelisMenten(prefix="menten")
logistic = LogisticSaturation(prefix="logistic")

X = pd.DataFrame(..., columns=["C1", "C2", "C3"])

coords = {"menten_channels": ["C1", "C2"], "logistic_channels": ["C3"]}
with pm.Model(coords=coords) as model:
    # Not hard coded to channel
    menten_contribution = menten.apply(x=X.loc[:, ["C1", "C2"]].to_numpy(), dim_name="menten_channels")
    logistic_contribution = logistic.apply(x=X.loc[:, ["C3"]].to_numpy(), dim_name="logistic_channels")

# This is where the prefix would no longer clash!
assert "menten_lam" in model
assert "logistic_lam" in model

Reply:
cool!